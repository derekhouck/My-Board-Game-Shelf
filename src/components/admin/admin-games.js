import React from 'react';
import requiresLogin from '../requires-login';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import { deleteGame, fetchGames } from '../../actions/games';

import Button from '../button';
import Loading from '../loading';
import Table from '../table';

export class AdminGames extends React.Component {
  state = {
    isLoading: false
  }

  componentDidMount() {
    const { dispatch, isAdmin } = this.props;
    if (isAdmin) {
      this.setState({ isLoading: true });
      dispatch(fetchGames()).then(() => this.setState({ isLoading: false }));
    }
  }

  handleDeleteButtonBlick(game) {
    const { dispatch } = this.props;
    this.setState({ isLoading: true });
    dispatch(deleteGame(game)).then(() => this.setState({ isLoading: false }));
  }

  render() {
    const { games, isAdmin } = this.props;
    const { isLoading } = this.state;

    const gamesTable = (
      <Table>
        <thead>
          <tr>
            <th>Title</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          {games.map(game => (
            <tr key={game.id}>
              <td>{game.title}</td>
              <td className="btn-group">
                <Link to={`/games/${game.id}/edit`}>
                  <Button label='Edit' small />
                </Link>
                <Button
                  className='btn--remove'
                  label='Delete'
                  onClick={() => this.handleDeleteButtonBlick(game)}
                  small
                />
              </td>
            </tr>
          ))}
        </tbody>
      </Table>
    );

    return isAdmin
      ? (
        <section>
          <h2>Games</h2>
          {isLoading ? <Loading /> : gamesTable}
        </section>
      )
      : <Redirect to="/" />;
  }
}

AdminGames.defaultProps = {
  games: [],
  isAdmin: false,
};

const mapStateToProps = state => ({
  games: state.games.games,
  isAdmin: state.auth.currentUser.admin,
});

export default requiresLogin()(connect(mapStateToProps)(AdminGames));