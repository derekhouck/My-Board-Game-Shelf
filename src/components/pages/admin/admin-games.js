import React from 'react';
import { requiresAdmin } from '../../helpers/requiresAdmin';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { fetchAdminGames } from '../../../actions/admin';
import { deleteGame } from '../../../actions/games';

import Button from '../../button';
import Loading from '../../loading';
import StatusIndicator from '../../atoms/status-indicator';
import Table from '../../table';

export class AdminGames extends React.Component {
  state = {
    filters: {
      status: ''
    },
    isLoading: false
  }

  componentDidMount() {
    const { dispatch } = this.props;
    dispatch(fetchAdminGames());
  }

  handleDeleteButtonBlick(game) {
    const { dispatch } = this.props;
    this.setState({ isLoading: true });
    dispatch(deleteGame(game)).then(() => this.setState({ isLoading: false }));
  }

  render() {
    const { error, games, loading } = this.props;
    const { filters, isLoading } = this.state;
    const addColor = status =>
      status === 'approved'
        ? 'green'
        : status === 'rejected'
          ? 'red'
          : 'yellow';
    const filteredGames = filters.status
      ? games.filter(game => game.status === filters.status)
      : games;

    const gamesTable = (
      <Table headings={['Title', 'Status', 'Actions']}>
        <tbody>
          {filteredGames.map(game => (
            <tr key={game.id}>
              <td>{game.title}</td>
              <td>
                <StatusIndicator
                  color={addColor(game.status)}
                >
                  {game.status}
                </StatusIndicator>
              </td>
              <td className="btn-group">
                <Link to={`/games/${game.id}/edit`}>
                  <Button label='Edit' small />
                </Link>
                <Button
                  className='btn--remove'
                  label='Delete'
                  onClick={() => this.handleDeleteButtonBlick(game)}
                  small
                />
              </td>
            </tr>
          ))}
        </tbody>
      </Table>
    );

    return (
      <section>
        <h2>Games</h2>
        {error && <StatusIndicator color="red">{`${error.status} ${error.name}: ${error.message}`}</StatusIndicator>}
        <div>
          Status:
            <select onChange={e => this.setState({ filters: { status: e.target.value } })}>
            <option value="">any</option>
            <option>pending</option>
            <option>approved</option>
            <option>rejected</option>
          </select>
        </div>
        {isLoading || loading ? <Loading /> : gamesTable}
      </section>
    );
  }
}

AdminGames.defaultProps = {
  error: null,
  games: [],
  loading: false,
};

const mapStateToProps = state => ({
  error: state.admin.error,
  games: state.admin.games,
  loading: state.admin.loading,
});

export default requiresAdmin()(connect(mapStateToProps)(AdminGames));