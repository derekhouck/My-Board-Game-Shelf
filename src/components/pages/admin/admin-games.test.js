import React from 'react';
import { shallow } from 'enzyme';
import { AdminGames } from './admin-games';

describe('<AdminGames />', function () {
  const dispatch = jest.fn().mockResolvedValue();

  it('renders a Games heading when admin', () => {
    const wrapper = shallow(
      <AdminGames dispatch={dispatch} />
    );
    expect(wrapper).toMatchSnapshot();
    expect(dispatch).toHaveBeenCalled();
    expect(wrapper.find('h2').text()).toEqual('Games');
  });

  it('renders each game title', () => {
    const games = [
      { id: 1, title: 'Game One' },
      { id: 2, title: 'Game Two' }
    ];
    const wrapper = shallow(
      <AdminGames
        dispatch={dispatch}
        games={games}
      />
    );
    wrapper.setState({ isLoading: false });
    games.forEach(game =>
      expect(wrapper.contains(game.title)).toBe(true)
    );
  });

  it('filters by status', () => {
    const games = [
      {
        id: 1,
        status: 'pending',
        title: 'Game One'
      },
      {
        id: 2,
        status: 'approved',
        title: 'Game Two'
      }
    ];
    const wrapper = shallow(
      <AdminGames
        dispatch={dispatch}
        games={games}
      />
    );

    wrapper.setState({
      filters: { status: 'approved' },
      isLoading: false,
    })
    expect(wrapper.contains(games[0].title)).toBe(false);
    expect(wrapper.contains(games[1].title)).toBe(true);
  });

  it('filters by name', () => {
    const games = [
      { id: 1, title: 'Game One' },
      { id: 2, title: 'Game Two' }
    ];
    const wrapper = shallow(
      <AdminGames
        dispatch={dispatch}
        games={games}
      />
    );
    wrapper.setState({
      filters: { name: 'one' },
      isLoading: false,
    });
    expect(wrapper.contains(games[0].title)).toBe(true);
    expect(wrapper.contains(games[1].title)).toBe(false);
  });
});