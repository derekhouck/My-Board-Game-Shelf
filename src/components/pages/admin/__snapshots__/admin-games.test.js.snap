// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminGames /> redirects when not admin 1`] = `ShallowWrapper {}`;

exports[`<AdminGames /> renders a Games heading when admin 1`] = `ShallowWrapper {}`;
